File: server/src/loaders.ts
import {
  JSONLoader,
  JSONLinesLoader
} from 'langchain/document_loaders/fs/json';

import { TextLoader } from 'langchain/document_loaders/fs/text';
import { CSVLoader } from 'langchain/document_loaders/fs/csv';
import { PDFLoader } from 'langchain/document_loaders/fs/pdf';
import { DocxLoader } from 'langchain/document_loaders/fs/docx';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { Document } from 'langchain/dist/document';

interface DocumentLoaders {
  [key: string]: (file: Blob) => Promise<any>;
}

const getFileExtension = (filename: string): string => {
  return filename
    .slice(((filename.lastIndexOf('.') - 1) >>> 0) + 2)
    .toLowerCase();
};

const getBlobFromFile = (file: Express.Multer.File): Blob => {
  const { buffer } = file;
  const type = file.mimetype;
  const blob = new Blob([buffer], { type });
  return blob;
};

const textSplitter = new RecursiveCharacterTextSplitter({
  chunkSize: 1000,
  chunkOverlap: 0
});

const loaders: DocumentLoaders = {
  json: async (file: Blob) => new JSONLoader(file, '/texts'),
  jsonl: async (file: Blob) => new JSONLinesLoader(file, '/html'),
  txt: async (file: Blob) => new TextLoader(file),
  tsx: async (file: Blob) => new TextLoader(file),
  ts: async (file: Blob) => new TextLoader(file),
  css: async (file: Blob) => new TextLoader(file),
  csv: async (file: Blob) => new CSVLoader(file),
  docx: async (file: Blob) => new DocxLoader(file),
  pdf: async (file: Blob) =>
    new PDFLoader(file, {
      splitPages: false
    })
};

export const loader = async (
  file: Express.Multer.File
): Promise<{
  documents: Document[];
  errors: { error: string; item: string }[];
}> => {
  const documents: Document[] = [];
  const errors: { error: string; item: string }[] = [];

  const blob = getBlobFromFile(file);
  const extension = getFileExtension(file.originalname);
  const loadFn = loaders[extension];

  if (!loadFn) {
    errors.push({
      error: `Unsupported file extension: ${extension}`,
      item: file.originalname
    });
  } else {
    const fileLoader = await loadFn(blob);
    const docs: Document[] = await fileLoader.loadAndSplit(textSplitter);
    const customMetadata = { filename: file.originalname };

    documents.push(
      ...docs.map((doc: Document) => {
        return {
          ...doc,
          metadata: {
            ...doc.metadata,
            ...customMetadata
          }
        };
      })
    );
  }

  return { documents, errors };
};

File: pyServer/server/services/loaders.py
from werkzeug.datastructures import FileStorage
from langchain.docstore.document import Document
from langchain.document_loaders import (
    PyPDFLoader,
    JSONLoader,
    TextLoader,
    CSVLoader,
    Docx2txtLoader,
)
from langchain.text_splitter import RecursiveCharacterTextSplitter
from typing import Union, List, Optional

ALLOWED_EXTENSIONS = {
    "json",
    "txt",
    "pdf",
    "tsx",
    "ts",
    "js",
    "css",
    "csv",
    "docx",
}


class LoaderError:
    error: str
    item: str

    def __init__(self, error: str, item: str):
        self.error = error
        self.item = item


class LoaderResult:
    documents: List[Document]
    errors: List[LoaderError]

    def __init__(
        self, documents: List[Document], errors: Optional[List[LoaderError]] = None
    ):
        self.documents = documents
        self.errors = errors if errors is not None else []


text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=100)

loaders = {
    "json": lambda file: JSONLoader(file, "/texts"),
    "pdf": lambda file: PyPDFLoader(file),
    "txt": lambda file: TextLoader(file),
    "tsx": lambda file: TextLoader(file),
    "ts": lambda file: TextLoader(file),
    "js": lambda file: TextLoader(file),
    "css": lambda file: TextLoader(file),
    "csv": lambda file: CSVLoader(file),
    "docx": lambda file: Docx2txtLoader(file),
}


def get_file_extension(filename: str) -> str:
    return filename.rsplit(".", 1)[-1].lower()


def loader(file_path: str, filename: str) -> LoaderResult:
    if not filename or not file_path:
        return LoaderResult([], [])

    errors: List[LoaderError] = []

    extension = get_file_extension(filename)

    load_fn = loaders.get(extension)

    if not load_fn:
        print(f"==========================Unsupported file extension: {extension}")
        errors.append(
            (
                LoaderError(
                    error=f"Unsupported file extension: {extension}", item=filename
                )
            )
        )

        return LoaderResult([], errors)

    file_loader = load_fn(file_path)

    documents: List[Document] = file_loader.load_and_split(text_splitter)

    for doc in documents:
        metadata = doc.metadata
        metadata["filename"] = filename

    return LoaderResult(documents, errors)
